File: app/src/main/java/com/mgaetan89/showsrage/fragment/AddShowOptionsFragment.java
package com.mgaetan89.showsrage.fragment;

import android.app.Dialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.FragmentActivity;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.SwitchCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.Spinner;

import com.mgaetan89.showsrage.Constants;
import com.mgaetan89.showsrage.R;
import com.mgaetan89.showsrage.activity.BaseActivity;
import com.mgaetan89.showsrage.adapter.RootDirectoriesAdapter;
import com.mgaetan89.showsrage.helper.GenericCallback;
import com.mgaetan89.showsrage.model.GenericResponse;
import com.mgaetan89.showsrage.network.SickRageApi;

import java.util.Set;

import retrofit.Callback;
import retrofit.client.Response;

public class AddShowOptionsFragment extends DialogFragment implements DialogInterface.OnClickListener {
	@Nullable
	private Spinner allowedQuality = null;

	@Nullable
	private SwitchCompat anime = null;

	@Nullable
	private Spinner language = null;

	@Nullable
	private Spinner preferredQuality = null;

	@Nullable
	private Spinner rootDirectory = null;

	@Nullable
	private Spinner status = null;

	@Nullable
	private SwitchCompat subtitles = null;

	public AddShowOptionsFragment() {
	}

	@Override
	public void onClick(DialogInterface dialog, int which) {
		int indexerId = this.getArguments().getInt(Constants.Bundle.INDEXER_ID, 0);

		if (indexerId <= 0) {
			return;
		}

		String allowedQuality = this.getAllowedQuality(this.allowedQuality);
		int anime = (this.anime != null && this.anime.isChecked()) ? 1 : 0;
		String language = this.getLanguage(this.language);
		String location = getLocation(this.rootDirectory);
		String preferredQuality = this.getPreferredQuality(this.preferredQuality);
		String status = this.getStatus(this.status);
		int subtitles = (this.subtitles != null && this.subtitles.isChecked()) ? 1 : 0;

		Callback<GenericResponse> callback = new AddShowCallback(this.getActivity());

		SickRageApi.Companion.getInstance().getServices()
				.addNewShow(indexerId, preferredQuality, allowedQuality, status, language, anime, subtitles, location, callback);
	}

	@NonNull
	@Override
	public Dialog onCreateDialog(Bundle savedInstanceState) {
		View view = LayoutInflater.from(this.getActivity()).inflate(R.layout.fragment_add_show_options, null);

		if (view != null) {
			this.allowedQuality = (Spinner) view.findViewById(R.id.allowed_quality);
			this.anime = (SwitchCompat) view.findViewById(R.id.anime);
			this.language = (Spinner) view.findViewById(R.id.language);
			this.preferredQuality = (Spinner) view.findViewById(R.id.preferred_quality);
			this.status = (Spinner) view.findViewById(R.id.status);
			this.subtitles = (SwitchCompat) view.findViewById(R.id.subtitles);

			LinearLayout rootDirectoryLayout = (LinearLayout) view.findViewById(R.id.root_directory_layout);

			if (rootDirectoryLayout != null) {
				SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this.getContext());
				Set<String> rootDirectories = preferences.getStringSet(Constants.Preferences.Fields.ROOT_DIRS, null);

				if (rootDirectories == null || rootDirectories.size() < 2) {
					rootDirectoryLayout.setVisibility(View.GONE);
				} else {
					this.rootDirectory = (Spinner) view.findViewById(R.id.root_directory);

					if (this.rootDirectory != null) {
						this.rootDirectory.setAdapter(new RootDirectoriesAdapter(this.getContext(), rootDirectories));
					}

					rootDirectoryLayout.setVisibility(View.VISIBLE);
				}
			}
		}

		AlertDialog.Builder builder = new AlertDialog.Builder(this.getActivity());
		builder.setTitle(R.string.add_show);
		builder.setView(view);
		builder.setPositiveButton(R.string.add, this);
		builder.setNegativeButton(R.string.cancel, null);

		return builder.show();
	}

	@Override
	public void onDestroyView() {
		this.allowedQuality = null;
		this.anime = null;
		this.language = null;
		this.preferredQuality = null;
		this.rootDirectory = null;
		this.status = null;
		this.subtitles = null;

		super.onDestroyView();
	}

	@Nullable
	/* package */ String getAllowedQuality(@Nullable Spinner allowedQualitySpinner) {
		int allowedQualityIndex = 0;

		if (allowedQualitySpinner != null) {
			// Skip the "Ignore" first item
			allowedQualityIndex = allowedQualitySpinner.getSelectedItemPosition() - 1;
		}

		if (allowedQualityIndex >= 0) {
			String qualities[] = this.getResources().getStringArray(R.array.allowed_qualities_keys);

			if (allowedQualityIndex < qualities.length) {
				return qualities[allowedQualityIndex];
			}
		}

		return null;
	}

	@Nullable
	/* package */ String getLanguage(@Nullable Spinner languageSpinner) {
		int languageIndex = 0;

		if (languageSpinner != null) {
			languageIndex = languageSpinner.getSelectedItemPosition();
		}

		if (languageIndex >= 0) {
			String languages[] = this.getResources().getStringArray(R.array.languages_keys);

			if (languageIndex < languages.length) {
				return languages[languageIndex];
			}
		}

		return null;
	}

	@Nullable
	/* package */ static String getLocation(@Nullable Spinner rootDirectorySpinner) {
		if (rootDirectorySpinner != null) {
			Object rootDirectory = rootDirectorySpinner.getSelectedItem();

			if (rootDirectory != null) {
				return rootDirectory.toString();
			}
		}

		return null;
	}

	@Nullable
	/* package */ String getPreferredQuality(@Nullable Spinner preferredQualitySpinner) {
		int preferredQualityIndex = 0;

		if (preferredQualitySpinner != null) {
			// Skip the "Ignore" first item
			preferredQualityIndex = preferredQualitySpinner.getSelectedItemPosition() - 1;
		}

		if (preferredQualityIndex >= 0) {
			String qualities[] = this.getResources().getStringArray(R.array.preferred_qualities_keys);

			if (preferredQualityIndex < qualities.length) {
				return qualities[preferredQualityIndex];
			}
		}

		return null;
	}

	@Nullable
	/* package */ String getStatus(@Nullable Spinner statusSpinner) {
		int statusIndex = 0;

		if (statusSpinner != null) {
			statusIndex = statusSpinner.getSelectedItemPosition();
		}

		if (statusIndex >= 0) {
			String statuses[] = this.getResources().getStringArray(R.array.status_keys);

			if (statusIndex < statuses.length) {
				return statuses[statusIndex];
			}
		}

		return null;
	}

	private static class AddShowCallback extends GenericCallback {
		private AddShowCallback(FragmentActivity activity) {
			super(activity);
		}

		@Override
		public void success(GenericResponse genericResponse, Response response) {
			super.success(genericResponse, response);

			FragmentActivity activity = this.getActivity();

			if (activity != null) {
				Intent intent = new Intent(activity, BaseActivity.class);

				activity.startActivity(intent);
			}
		}
	}
}

File: app/src/main/kotlin/com/mgaetan89/showsrage/fragment/AddShowOptionsFragment.kt
package com.mgaetan89.showsrage.fragment

import android.app.Dialog
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.support.v4.app.FragmentActivity
import android.support.v7.app.AlertDialog
import android.support.v7.widget.SwitchCompat
import android.view.View
import android.widget.LinearLayout
import android.widget.Spinner
import com.mgaetan89.showsrage.Constants
import com.mgaetan89.showsrage.R
import com.mgaetan89.showsrage.activity.MainActivity
import com.mgaetan89.showsrage.adapter.RootDirectoriesAdapter
import com.mgaetan89.showsrage.extension.getRootDirs
import com.mgaetan89.showsrage.extension.inflate
import com.mgaetan89.showsrage.extension.toInt
import com.mgaetan89.showsrage.helper.GenericCallback
import com.mgaetan89.showsrage.model.GenericResponse
import com.mgaetan89.showsrage.model.RootDir
import com.mgaetan89.showsrage.network.SickRageApi
import io.realm.Realm
import retrofit.client.Response

class AddShowOptionsFragment : DialogFragment(), DialogInterface.OnClickListener {
	private var allowedQuality: Spinner? = null
	private var anime: SwitchCompat? = null
	private var futureStatus: Spinner? = null
	private var language: Spinner? = null
	private var preferredQuality: Spinner? = null
	private var rootDirectory: Spinner? = null
	private var seasonFolders: SwitchCompat? = null
	private var status: Spinner? = null
	private var subtitles: SwitchCompat? = null

	override fun onClick(dialog: DialogInterface?, which: Int) {
		val activity = this.activity ?: return
		val indexerId = this.arguments?.getInt(Constants.Bundle.INDEXER_ID) ?: return

		if (indexerId <= 0) {
			return
		}

		val allowedQuality = this.getAllowedQuality(this.allowedQuality)
		val anime = this.anime?.isChecked.toInt()
		val callback = AddShowCallback(activity)
		val futureStatus = this.getStatus(this.futureStatus)
		val language = this.getLanguage(this.language)
		val location = getLocation(this.rootDirectory)
		val preferredQuality = this.getPreferredQuality(this.preferredQuality)
		val seasonFolders = this.seasonFolders?.isChecked.toInt()
		val status = this.getStatus(this.status)
		val subtitles = this.subtitles?.isChecked.toInt()

		SickRageApi.instance.services?.addNewShow(indexerId, preferredQuality, allowedQuality, status, futureStatus, language, anime, seasonFolders, subtitles, location, callback)
	}

	override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
		val context = this.context ?: return super.onCreateDialog(savedInstanceState)
		val view = context.inflate(R.layout.fragment_add_show_options)

		this.allowedQuality = view.findViewById(R.id.allowed_quality) as Spinner?
		this.anime = view.findViewById(R.id.anime) as SwitchCompat?
		this.futureStatus = view.findViewById(R.id.future_status) as Spinner?
		this.language = view.findViewById(R.id.language) as Spinner?
		this.preferredQuality = view.findViewById(R.id.preferred_quality) as Spinner?
		this.seasonFolders = view.findViewById(R.id.season_folders) as SwitchCompat?
		this.status = view.findViewById(R.id.status) as Spinner?
		this.subtitles = view.findViewById(R.id.subtitles) as SwitchCompat?

		val rootDirectoryLayout = view.findViewById(R.id.root_directory_layout) as LinearLayout?

		if (rootDirectoryLayout != null) {
			val realm = Realm.getDefaultInstance()
			val rootDirectories = realm.getRootDirs()

			if (rootDirectories.size < 2) {
				rootDirectoryLayout.visibility = View.GONE
			} else {
				this.rootDirectory = view.findViewById(R.id.root_directory) as Spinner?

				if (this.rootDirectory != null) {
					this.rootDirectory!!.adapter = RootDirectoriesAdapter(rootDirectories)
				}

				rootDirectoryLayout.visibility = View.VISIBLE
			}

			realm.close()
		}

		val builder = AlertDialog.Builder(context)
		builder.setTitle(R.string.add_show)
		builder.setView(view)
		builder.setPositiveButton(R.string.add, this)
		builder.setNegativeButton(R.string.cancel, null)

		return builder.show()
	}

	override fun onDestroyView() {
		this.allowedQuality = null
		this.anime = null
		this.futureStatus = null
		this.language = null
		this.preferredQuality = null
		this.rootDirectory = null
		this.seasonFolders = null
		this.status = null
		this.subtitles = null

		super.onDestroyView()
	}

	fun getAllowedQuality(allowedQualitySpinner: Spinner?): String? {
		val allowedQualityIndex = (allowedQualitySpinner?.selectedItemPosition ?: 1) - 1

		if (allowedQualityIndex >= 0) {
			val qualities = this.resources.getStringArray(R.array.allowed_qualities_keys)

			if (allowedQualityIndex < qualities.size) {
				return qualities[allowedQualityIndex]
			}
		}

		return null
	}

	fun getLanguage(languageSpinner: Spinner?): String? {
		val languageIndex = languageSpinner?.selectedItemPosition ?: 0

		if (languageIndex >= 0) {
			val languages = this.resources.getStringArray(R.array.languages_keys)

			if (languageIndex < languages.size) {
				return languages[languageIndex]
			}
		}

		return null
	}

	fun getPreferredQuality(preferredQualitySpinner: Spinner?): String? {
		val preferredQualityIndex = (preferredQualitySpinner?.selectedItemPosition ?: 1) - 1

		if (preferredQualityIndex >= 0) {
			val qualities = this.resources.getStringArray(R.array.preferred_qualities_keys)

			if (preferredQualityIndex < qualities.size) {
				return qualities[preferredQualityIndex]
			}
		}

		return null
	}

	fun getStatus(statusSpinner: Spinner?): String? {
		val statusIndex = statusSpinner?.selectedItemPosition ?: 0

		if (statusIndex >= 0) {
			val statuses = this.resources.getStringArray(R.array.status_keys)

			if (statusIndex < statuses.size) {
				return statuses[statusIndex]
			}
		}

		return null
	}

	companion object {
		fun getLocation(rootDirectorySpinner: Spinner?): String? {
			val selectedItem = rootDirectorySpinner?.selectedItem

			if (selectedItem is RootDir) {
				return selectedItem.location
			}

			return null
		}

		fun newInstance(indexerId: Int) = AddShowOptionsFragment().apply {
			this.arguments = Bundle().apply {
				this.putInt(Constants.Bundle.INDEXER_ID, indexerId)
			}
		}
	}

	private class AddShowCallback(activity: FragmentActivity) : GenericCallback(activity) {
		override fun success(genericResponse: GenericResponse?, response: Response?) {
			super.success(genericResponse, response)

			val activity = this.getActivity() ?: return
			val intent = Intent(activity, MainActivity::class.java)

			activity.startActivity(intent)
		}
	}
}
